### ISSUE TITLE
Regression: Incorrect video scaling on wayland

### ISSUE THREAD

**Issue by cherti** [mpv-player/mpv#9426]
### Important Information

- mpv version: 0.34 (tested with master as well)
- Linux Distribution and Version: Arch Linux, current™
- Source of the mpv binary: - https://archlinux.org/packages/community/x86_64/mpv/
                                            - https://github.com/archlinux/svntogit-community/blob/packages/mpv/trunk/PKGBUILD
- If known which version of mpv introduced the problem: 0.34
- Window Manager and version: Wayland: sway 1.6.1
- GPU driver and version: Intel-GPU in Linux 5.14.16
- Possible screenshot or video of visual glitches: Video starts incorrectly scaled

### Reproduction steps

`--no-config` didn't help, hence the reproduction steps below.
Unfortunately, on this multi-monitor-setup, the problem only exists on one of the three monitors in question (all of them differently scaled due to different DPI). The affected screen is the one that is unscaled, however it is unclear to me if that is coincidence or meaningful to the regression.

Assumed the video initially gets opened on the one affected screen (screenshots further down under expected and actual behavior):

1. get a video file (tested with VP_9, mp4, Youtube-Streaming)
2. call `mpv <videofile/videolink>
3. observe a video that is (usually, see 5) way too large for the window it's played in
4. full-screen and de-fullscreen -> issue's gone
5. move video to another screen -> problem returns, this time video gets too small
6. full-screen and de-fullscreen -> issue's gone
7. move back to screen where it's initially opened: original issue arises again

4 and 6 can also be de-fullscreen and re-fullscreen, same behavior
5's exact behavior might be related to different scalings on different screens

Given that only some screens are affected combined with 7. might maybe point to mpv picking the wrong scaling factors for the display in use, but that's guesswork on my side given the evidence and might be entirely incorrect.

### Bisection

I bisected the code and the problem seems to be introduced by a02901cae77c86fb9de997a418296d1fc0e3eada, which seems plausible given the commit message reads "wayland: fix wl_surface_set_buffer_scale usage". Unfortunately I don't see myself capable of fixing this issue right away, so I'm afraid I have to recourse to merely reporting.

### Log file

[mpvlog.txt](https://github.com/mpv-player/mpv/files/7496721/mpvlog.txt)

### Expected behavior

Video always has the correct size so that it fits the window:
![good](https://user-images.githubusercontent.com/6452205/140737414-3fee9f6f-8f62-40c8-ad26-fcdb1b6a6a19.png)

### Actual behavior

Video is incorrectly scaled upon opening:
![bad](https://user-images.githubusercontent.com/6452205/140737454-a3ea8b28-321b-472d-8658-9de28c370d99.png)

**Comment by Dudemanguy** [comment]
mpv has a bug but actually so does this version of sway (it's fixed in master). So it's an unfortunate collusion of things causing this behavior. I'm assuming that the mpv in your example is floating. In that case, it has this sway bug (https://github.com/swaywm/sway/issues/6080) with regards to fullscreening a floating window. What happened was that it incorrectly intersected the geometry of another monitor and thus caused an erroneous surface entrance event.

As for the mpv bug, when I refactored the surface scaling handling, I added a conditional in there that looks very suspicious and erroneous (and indeed getting rid of it seems to be correct). I'll kick it around a bit more to be sure but I think with that small fix + using sway from master it should work correctly again.

**Comment by cherti** [comment]
Thanks for the feedback!

> I'm assuming that the mpv window in your example is floating.

That's incorrect, the window is tiled, not floating. Does that change your conclusion?

**Comment by Dudemanguy** [comment]
>Does that change your conclusion?

Unfortunately yes. As far as I know, there was never a problem in tiled mode (I can't reproduce but my setup is nowhere near as complex as yours).

**Comment by quietvoid** [comment]
I don't have an issue either. Sway, multiple monitors but none are scaled.
My tile is scaled properly:
![2021-11-08T11:09:54,138296801-05:00](https://user-images.githubusercontent.com/39477805/140777082-e44bf3a6-35e5-4844-838b-537b61023260.png)

**Comment by cherti** [comment]
> but none are scaled.

I think that explains why you don't see the problem, it very much looks like a scaling problem from my testing.

@Dudemanguy checked sway master, and indeed I cannot reproduce the issue anymore with that. (unfortunately, sway-master breaks wayland-electron apps, so sigh… seems I have to pick my poison…). Should I close here or is there any reason to keep it open for longer?

(and do you, by any chance have the commit handy that fixed that in sway? (Don't start looking if you don't, it's not worth wasting your time :)) maybe I can cherrypick and run my own patchwork there… or just pin mpv at 0.33 for the moment until sway 1.6.2 which then hopefully doesn't break those electron-apps. anyways...)

**Comment by Dudemanguy** [comment]
Hmm, I'm not sure what what have fixed this. My commit should have only affected floating windows so I don't think it is that. But yeah, I think this can be closed now.

**Comment by cherti** [comment]
alright, thanks tremendously for the help! :)

**Comment by cherti** [comment]
@Dudemanguy btw, if you have anything you want to have tested, hit me up with it. :)

**Comment by cherti** [comment]
uhm, the problem is still there with the recent upgrade to sway 1.7 now… any idea what the reason could be? I have tried testing master, but that currently doesn't work for reasons that are unrelated to this problem and I can't investigate deeply right now.

sway: 1.7
wlroots: 0.15
mpv: 0.34.1

**Comment by Dudemanguy** [comment]
I'm on the new sway/wlroots and I can't reproduce. Any chance #9766 fixes it?

**Comment by cherti** [comment]
unfortunately not. Do you have differently scaled displays? Because it looks to me, that when moving mpv between screens, it carries the scaling factor of the previous screen with it, and only updates that on the first resize.
So when I have a fullscreened mpv, I used to be able to move that one between screens keeping it fullsized, whereas it now looks like it would take the previous scaling with it.

When I press i for the infos on the vids, the scaled resolution differs, though from the two screens, and when I move a fullscreened window from a screen with scaling 1 to one with scaling 1.5 it will be scaled to small, and when I move it back immediately it will be scaled too large. So could it be that either mpv obtains the wrong screen's scaling factor or sway returns the wrong screen scaling factor?

**Comment by Dudemanguy** [comment]
I don't normally use multiple displays but I could set it up and test it out later. On master, mpv is supposed to update scaling whenever it receives a surface event at least.

>So could it be that either mpv obtains the wrong screen's scaling factor or sway returns the wrong screen scaling factor?

That seems possible. Could you post an updated log file against master?

**Comment by cherti** [comment]
what's the preferred way of obtaining the log? simply stdout and stderr or is there something more illustrative that could help?

**Comment by Dudemanguy** [comment]
`--log-file=mpv.log`

**Comment by Dudemanguy** [comment]
Well according to the log, the first entrance event we get is:
`Surface entered output Dell Inc. DELL U2515H (0x2a), scale = 1`
Then:
`Surface entered output Unknown 0x058B (0x2d), scale = 2`
Finally:
`Surface entered output Dell Inc. DELL U2515H (0x2a), scale = 1`

I assume this is backwards.

**Comment by cherti** [comment]
That does look correct, yes.

Interesting effect: when testing the above I had s1 disabled and reenabled (because on older firefox versions there was a bug with context menus that got kinda hotfixed-ish by that). Now I tried without this hotfix and while 2-5 stay the same, the video in 1 now gets opened at correct window size (which is window width, effectively, because the window is higher than wide). Not sure if that is helpful in any way, but I try at least to get as much info as I can to at least help on that part. :)

**Comment by Dudemanguy** [comment]
Okay so I tried to reproduce this on a dual screen setup with different scale factors (set a variety of combinations), but I could not. The tiled window was always the correct size no matter where I opened the video or how I moved it around.

**Comment by cherti** [comment]
hm, and you are also on sway 1.7, right?

**Comment by Dudemanguy** [comment]
Correct.

**Comment by cherti** [comment]
I've tinkered a bit and I might have some more pointers (or I might not, not totally sure).
When I use the following configuration:

    output DP-3 res 2560x1440 scale 1
    output eDP-1 res 2560x1440 scale 1
    output DP-4 res 1920x1200 scale 1

Everything works fine. However, when I use the following configuration:

    output DP-3 res 2560x1440 scale 1
    output eDP-1 res 2560x1440 scale 2
    output DP-4 res 1920x1200 scale 1

I observe the following:

Moving from DP-4 to DP-3: everything is scaled as expected when moving between screens.
Same for DP-3 -> DP-4, DP-4 -> eDP1, eDP-1 -> DP-4

However, when I move
eDP1 -> DP-3: the video on DP-3 is, when landing on that screen, always* exactly* twice as big as it should be.
conversely, when I move it
DP-3 -> eDP1: the video now on eDP1 is exactly* half as big as it should be (centered and left and right is exactly half the video-width of unoccupied space).

* as far as I can tell with measuring with a ruler

I can reliably reproduce that behavior with having that video on pause and moving that window over the screens. Consequentially, moving eDP1 -> DP-4 -> DP-3 everything behaves according to expectations and the other way around as well.
It's just directly between DP-3 and eDP1 that this issue occurs.

Does that point to anything, possibly?

**Comment by Dudemanguy** [comment]
>Does that point to anything, possibly?

Yes actually. I think I know the issue. It's not doing a resize when it always should be doing one. It would also explain why I couldn't reproduce it (because my monitors were different sizes and thus triggered a resize). Does this patch fix it?
```
diff --git a/video/out/wayland_common.c b/video/out/wayland_common.c
index 1f176aaad3..a9b2341340 100644
--- a/video/out/wayland_common.c
+++ b/video/out/wayland_common.c
@@ -700,6 +700,7 @@ static void surface_handle_enter(void *data, struct wl_surface *wl_surface,
     if (wl->scaling != wl->current_output->scale) {
         set_surface_scaling(wl);
         spawn_cursor(wl);
+        force_resize = true;
         wl->pending_vo_events |= VO_EVENT_DPI;
     }
```

**Comment by cherti** [comment]
Bingo! Thanks so much! can't reproduce in any direction anymore with that patch.

**Comment by Dudemanguy** [comment]
It's actually weird that merely building sway from master ever resolved this issue (or at least appeared to resolve it).

**Comment by cherti** [comment]
agreed… I tried to actually get an old version of mpv running to verify if I could actually reproduce any of that and didn't post incorrect results (although I'm fairly sure I checked before posting that), but unfortunately I think there was a move to newer dependencies or something so I couldn't actually get an older version of mpv to compile (and I admittedly didn't feel like experimentally downgrading dependencies to hunt for verification). Maybe something in combination of things triggered a resize in mpv for a different reason (possibly even a sway bug? otherwise it's unlikely it would be gone in 1.7. Or it just looked like the same bug and it were actually two different bugs just looking similar?

But yeah, it's a bit strange. I'm not excluding that I messed up something somewhere in the middle. But hey, at least it's properly fixed now!
